pkg/softtainter/node.go

```diff
@@ -0,0 +1,125 @@
+package softtainter
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// IsNodeSoftTainted checks if the node is already tainted
+// with the provided key and value and PreferNoSchedule effect.
+// If a match is found, it returns true.
+func IsNodeSoftTainted(node *v1.Node, taintKey, taintValue string) bool {
```

Comment:
[k8s.io/kubernetes/pkg/util/taints](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/taints/taints.go) has useful functions you could utilize here. E.g. `AddOrUpdateTaint`, `AddOrUpdateTaint` or `TaintExists`.

Reply:
I took `TaintExists`, `AddOrUpdateTaint` is less useful here since it will amend the node object node but here i'm using a patch operator to prevent conflicts.